rust/hyperlane-base/src/types/multisig.rs

```diff
@@ -2,105 +2,54 @@ use std::collections::{hash_map::Entry, HashMap};
-                        1 // length of 1
-                    }
-                };
-                // If we've hit a quorum, create a MultisigSignedCheckpoint
-                if signature_count >= self.threshold {
-                    if let Some(signed_checkpoints) = signed_checkpoints_per_root.get(&root) {
-                        let checkpoint = MultisigSignedCheckpoint::try_from(signed_checkpoints)?;
-                        debug!(checkpoint=?checkpoint, "Fetched multisig checkpoint");
-                        return Ok(Some(checkpoint));
-                    }
-                }
-            }
-        }
-
-        Ok(None)
-    }
-
-    /// Attempts to get the latest index with a quorum of signatures among validators.
+    // TODO: Return an error if we can't find a checkpoint
+    /// Attempts to get the latest checkpoint with a quorum of signatures among validators.
+    ///
     /// First iterates through the `latest_index` of each validator's checkpoint syncer,
     /// looking for the highest index that >= `threshold` validators have returned.
-    /// If there isn't a quorum found this way, each unique index from the highest -> lowest
-    /// is checked for a quorum of signed checkpoints using `fetch_checkpoint`.
-    /// Note it's possible for both strategies for finding the latest index to not find a quorum.
-    /// A more robust implementation should be considered in the future that indexes historical
-    /// checkpoint indices.
+    ///
+    /// Attempts to find a quorum of signed checkpoints from that index, iterating
+    /// backwards if unsuccessful, until the (optional) index is reached.
+    ///
+    /// Note it's possible to not find a quorum.
     #[instrument(err, skip(self))]
-    pub async fn latest_index(&self) -> Result<Option<u32>> {
+    pub async fn fetch_checkpoint_in_range(
+        &self,
+        validators: Vec<H256>,
+        threshold: usize,
+        maximum_index: u32,
+        minimum_index: u32,
+    ) -> Result<Option<MultisigSignedCheckpoint>> {
         // Get the latest_index from each validator's checkpoint syncer.
-        let mut latest_indices = Vec::with_capacity(self.checkpoint_syncers.len());
-        for checkpoint_syncer in self.checkpoint_syncers.values() {
-            // Gracefully handle errors getting the latest_index
-            if let Ok(Some(index)) = checkpoint_syncer.latest_index().await {
-                latest_indices.push(index);
+        let mut latest_indices = Vec::with_capacity(validators.len());
+        for validator in validators.iter() {
+            let addr = h256_to_h160(*validator);
+            if let Some(checkpoint_syncer) = self.checkpoint_syncers.get(&addr) {
+                // Gracefully handle errors getting the latest_index
+                if let Ok(Some(index)) = checkpoint_syncer.latest_index().await {
+                    latest_indices.push(index);
+                }
             }
```

Comment:
should we log if we get None from the `if let Some(checkpoint_syncer) = self.checkpoint_syncers.get(&addr) {` ?